src/txmempool.cpp

```diff
@@ -57,46 +57,75 @@ size_t CTxMemPoolEntry::GetTxSize() const
+    // already de-duplicated in case multiple outputs of ours are spent in one
+    // transaction).
+    // Always creates a new cache line entry.
+    vecEntries& new_cache_line = cache[update_it];
+    assert(new_cache_line.empty());
+    // next_idx_to_walk index keeps track of the elements that we've
+    // already expanded.
+    // Invariants:
+    // If index is < next_idx_to_walk, we've already walked it.
+    // If index is >= next_idx_to_walk, we need to walk it.
+    // If next_idx_to_walk >= new_cache_line.size(), we're finished.
+    size_t next_idx_to_walk = 0;
+    txiter next_it = update_it;
+    do {
+        for (const txiter& child_it : GetMemPoolChildren(next_it)) {
+            // N.B. children can also be grand children, so guarding
+            // here is neccessary
+            if (visited(child_it)) continue;
+            // if it exists in the cache, copy cached descendants
+            // - All elements in cache are not excluded, no need to check when adding.
+            // - All keys in the cache are already excluded, no need to add
+            //   `child_it` to the new_cache_line (but not all excluded keys
+            //   are in the cache yet!)
+            cacheMap::iterator cached_grand_children = cache.find(child_it);
+            if (cached_grand_children != cache.end()) {
+                for (const txiter& grand_child_it : cached_grand_children->second) {
+                    if (visited(grand_child_it)) continue;
+                    // place on the back and then swap into the next_idx_to_walk index
+                    // so we don't walk it ourselves (whoever put the grand
+                    // child in the cache must have already traversed this, so it must
+                    // not be excluded)
+                    new_cache_line.emplace_back(grand_child_it);
+                    std::swap(new_cache_line[next_idx_to_walk], new_cache_line.back());
+                    ++next_idx_to_walk;
                 }
-            } else if (!setAllDescendants.count(childEntry)) {
+            } else {
                 // Schedule for later processing
-                stageEntries.insert(childEntry);
+                // Element must not be excluded, or it would be in the cache already.
+                new_cache_line.emplace_back(child_it);
             }
         }
+
+        // finish loop here!
+        if (next_idx_to_walk >= new_cache_line.size()) break;
+
+        // Prepare for next iteration:
+        next_it = new_cache_line[next_idx_to_walk];
+        ++next_idx_to_walk;
+    } while (true);
+
+    // new_cache_line now contains all in-mempool descendants of update_it,
+    // compute updates now.
+    int64_t modify_size = 0;
+    CAmount modify_fee = 0;
+    int64_t modify_count = 0;
+    // All entries in the new_cache_line have
+    // already been checked against excluded list
+    for (txiter child_it : new_cache_line) {
```

Comment:
Ok so suppose vhashestoupdate is {X, Y}.

First we submit X to mempool and Y to mempool.

Then we reverse iterate and update Y, then X.

Previously, we scanned across all of the direct children in the mempool of Y then of X and updated the parent/child relationships. So Y would do only things already in the mempool, and X would see Y in this list, but exclude it in setExcluded as it was set properly when Y was put in the mempool.

In the new patch, we scanned across all of the direct children in the mempool of Y then of X and updated the parent/child relationships. So Y would do only things already in the mempool, and X would see Y in this list, but exclude it in mapMemPoolDescendants as it was set properly when Y was put in the mempool, and mapMemPoolDescendants will get a cacheline entry for everything in vHashesToUpdate; and because of iteration order, children will always be visible to parents.

When we actually call into UpdateForDescendants, we only update against things that are in the cachelines, and not the keys to those cachelines. This is because earlier when Y was accepted to the mempool it should have already modified X's fees.

Reply:
Gotcha, thanks for the laid-out explanation, I forgot that ancestor state X is well-modified at Y new mempool insertion. I was just confused by the sybillic comment "because any descendants in cache were added to the mempool after the transaction being updated and hence their state is already reflected in the parent state". In fact that's should point to `AcceptToMemoryPool` in validation's `UpdateMempoolForReorg` to enhance clarity.